File: typescript/examples/object-deep-merge.test.ts
// カード object-deep-merge の Contract を検証するテスト
import { toMerged } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

describe('object-deep-merge: toMerged', () => {
  it('ネストしたオブジェクトを再帰的にマージする', () => {
    const defaults = { retry: 3, log: { level: 'info', file: 'app.log' } };
    const overrides = { log: { level: 'debug' } };
    expect(toMerged(defaults, overrides)).toEqual({
      retry: 3,
      log: { level: 'debug', file: 'app.log' },
    });
  });

  it('target も source も変更しない（target は深くコピーされる）', () => {
    const target = { a: { x: 1 }, when: new Date(0) };
    const source = { a: { y: 2 } };
    const result = toMerged(target, source);
    expect(target).toEqual({ a: { x: 1 }, when: new Date(0) });
    expect(source).toEqual({ a: { y: 2 } });
    expect(result.a).not.toBe(target.a);
    expect(result.when).not.toBe(target.when);
  });

  it('配列はインデックスごとに上書きし、連結しない', () => {
    expect(toMerged({ a: [1, 2, 3] }, { a: [9] })).toEqual({ a: [9, 2, 3] });
    expect(toMerged({ a: [1] }, { a: [8, 9] })).toEqual({ a: [8, 9] });
    expect(toMerged({ a: [{ x: 1 }] }, { a: [{ y: 2 }] })).toEqual({ a: [{ x: 1, y: 2 }] });
  });

  it('source のプレーンオブジェクト・配列はコピーされ、それ以外のオブジェクトは参照が入る', () => {
    const source = { o: { z: 1 }, arr: [1], d: new Date(0) };
    const result = toMerged({ o: 1 }, source);
    expect(result.o).not.toBe(source.o);
    expect(result.arr).not.toBe(source.arr);
    expect(result.d).toBe(source.d);
  });

  it('source の undefined は定義済みの値を上書きせず、null は上書きする', () => {
    expect(toMerged({ a: 1, b: 2 }, { a: undefined, b: null })).toEqual({ a: 1, b: null });
  });

  it('片方だけがプレーンオブジェクト/配列なら source の値で置き換える', () => {
    expect(toMerged({ a: { x: 1 } }, { a: 5 })).toEqual({ a: 5 });
    expect(toMerged({ a: 5 }, { a: { x: 1 } })).toEqual({ a: { x: 1 } });
    expect(toMerged({ a: null }, { a: [1, 2, 3] })).toEqual({ a: [1, 2, 3] });
  });

  it('source の自身の列挙可能なキーだけを見て、__proto__ は無視する', () => {
    const source: Record<string, number> = Object.create({ inherited: 1 });
    source.own = 2;
    expect(toMerged({}, source)).toEqual({ own: 2 });

    const polluted = JSON.parse('{"__proto__": {"polluted": true}}') as Record<string, unknown>;
    const result = toMerged({}, polluted) as Record<string, unknown>;
    expect(result.polluted).toBeUndefined();
    expect(({} as Record<string, unknown>).polluted).toBeUndefined();
  });
});

File: python/examples/object-deep-merge_test.py
# カード object-deep-merge の Contract を検証するテスト
import pytest
from mergedeep import Strategy, merge


def test_merges_nested_dicts_recursively():
    """ネストした辞書を再帰的にマージする"""
    defaults = {"retry": 3, "log": {"level": "info", "file": "app.log"}}
    overrides = {"log": {"level": "debug"}}
    assert merge({}, defaults, overrides) == {"retry": 3, "log": {"level": "debug", "file": "app.log"}}


def test_mutates_and_returns_destination():
    """第 1 引数を破壊的に更新し、それ自身を返す。sources は変更しない"""
    dest = {"x": {"y": 1}}
    src = {"x": {"z": 2}}
    result = merge(dest, src)
    assert result is dest
    assert dest == {"x": {"y": 1, "z": 2}}
    assert src == {"x": {"z": 2}}


def test_empty_destination_keeps_sources_unchanged():
    """merge({}, a, b) の形なら a も b も変わらない"""
    a = {"log": {"level": "info"}, "tags": [1, 2, 3]}
    b = {"log": {"level": "debug"}, "tags": [9]}
    merge({}, a, b)
    assert a == {"log": {"level": "info"}, "tags": [1, 2, 3]}
    assert b == {"log": {"level": "debug"}, "tags": [9]}


def test_later_sources_win():
    """sources は左から順に適用され、後のものが勝つ"""
    assert merge({}, {"a": 1}, {"a": 2}, {"a": 3}) == {"a": 3}


def test_replace_strategy_replaces_non_mapping_values():
    """既定の REPLACE ではリストやスカラーは後の値で丸ごと置き換える（連結しない）"""
    assert merge({}, {"tags": [1, 2, 3]}, {"tags": [9]}) == {"tags": [9]}
    assert merge({}, {"a": {"b": 1}}, {"a": 5}) == {"a": 5}
    assert merge({}, {"a": 5}, {"a": {"b": 1}}) == {"a": {"b": 1}}


def test_values_are_deep_copied_from_sources():
    """sources の値は deepcopy され、返り値と参照を共有しない"""
    src = {"a": {"b": {"c": 1}}, "tags": [1]}
    result = merge({}, src)
    assert result == src
    assert result["a"] is not src["a"]
    assert result["a"]["b"] is not src["a"]["b"]
    assert result["tags"] is not src["tags"]


def test_none_overrides_existing_value():
    """sources の値が None でも上書きする"""
    assert merge({}, {"a": 1}, {"a": None}) == {"a": None}


def test_key_order_is_destination_then_new_keys():
    """キー順は destination の順の後に、新たに現れたキーがその順で並ぶ"""
    result = merge({}, {"b": 1, "a": 2}, {"c": 3, "a": 4})
    assert list(result) == ["b", "a", "c"]


def test_additive_strategy_concatenates_collections():
    """ADDITIVE では list / tuple / set を連結し、スカラーは置き換える"""
    assert merge({}, {"t": [1, 2]}, {"t": [3]}, strategy=Strategy.ADDITIVE) == {"t": [1, 2, 3]}
    assert merge({}, {"t": (1,)}, {"t": (2,)}, strategy=Strategy.ADDITIVE) == {"t": (1, 2)}
    assert merge({}, {"t": {1}}, {"t": {2}}, strategy=Strategy.ADDITIVE) == {"t": {1, 2}}
    assert merge({}, {"t": 1}, {"t": 2}, strategy=Strategy.ADDITIVE) == {"t": 2}


def test_typesafe_strategies_raise_on_type_mismatch():
    """TYPESAFE_REPLACE / TYPESAFE_ADDITIVE / TYPESAFE は型が異なると TypeError。REPLACE は投げない"""
    for strategy in (Strategy.TYPESAFE_REPLACE, Strategy.TYPESAFE_ADDITIVE, Strategy.TYPESAFE):
        with pytest.raises(TypeError):
            merge({}, {"t": 1}, {"t": "s"}, strategy=strategy)
    assert merge({}, {"t": 1}, {"t": 2}, strategy=Strategy.TYPESAFE_REPLACE) == {"t": 2}
    assert merge({}, {"t": 1}, {"t": "s"}) == {"t": "s"}


def test_no_sources_returns_destination_as_is():
    """sources が無ければ destination をそのまま返す"""
    assert merge({}) == {}


def test_shallow_merge_alternatives_replace_nested_dicts():
    """Alternatives: {**a, **b} と a | b はネストした辞書を丸ごと置き換える"""
    a = {"log": {"level": "info", "file": "app.log"}}
    b = {"log": {"level": "debug"}}
    assert {**a, **b} == {"log": {"level": "debug"}}
    assert a | b == {"log": {"level": "debug"}}
